R/derive_advs_params.R

```diff
@@ -0,0 +1,600 @@
+  assert_data_frame(dataset, required_vars = exprs(!!!by_vars, PARAMCD, AVAL))
+  assert_character_scalar(numerator_code)
+  assert_character_scalar(denominator_code)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_param_does_not_exist(dataset, set_values_to$PARAMCD)
+  assert_logical_scalar(constant_numerator)
+  assert_logical_scalar(constant_denominator)
+  filter <- assert_filter_cond(enexpr(filter), optional = TRUE)
+  assert_vars(constant_by_vars, optional = TRUE)
+  get_unit_expr <- assert_expr(enexpr(get_unit_expr), optional = TRUE)
+
+  ### Default formula with no units conversion applied ----
+
+  ratio_formula <- expr(
+    !!sym(paste0("AVAL.", numerator_code)) /
+      !!sym(paste0("AVAL.", denominator_code))
+  )
+
+  ### If `get_unit_expr` provided then check units and enable units conversion ----
+
+  if (!missing(get_unit_expr) && !is.null(get_unit_expr)) {
+    param_units <- list()
+
+    # Check if units are the same within each of two parameters
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
+        filter(PARAMCD == param) %>%
+        pull(`_unit`) %>%
+        unique()
+
+      if (length(units_found) != 1L) {
+        cli_abort(
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
+
+    # If the input parameters are measured in different units
+    # but are convertible from one to another (and this kind of conversion supported)
+    # then modify the formula in order to perform units conversion on the fly
+
+    if (param_units[[denominator_code]] != param_units[[numerator_code]]) {
+      # Find conversion factor for denominator
+      conv_factor <- get_conv_factor(
+        param_units[[denominator_code]],
+        param_units[[numerator_code]]
+      )
+
+      if (!is.na(conv_factor)) {
+        ratio_formula <- expr(
+          !!sym(paste0("AVAL.", numerator_code)) / (
+            !!sym(paste0("AVAL.", denominator_code)) * !!conv_factor
+          )
+        )
+
+        cli_alert_info(
+          "Unit conversion performed for {.val {denominator_code}}. Values converted from
+          {.val {param_units[[denominator_code]]}} to {.val {param_units[[numerator_code]]}}.",
+          wrap = TRUE
+        )
+      }
+    }
+  }
+
+  ### Identify constant parameters ----
+
+  parameters <- c(numerator_code, denominator_code)
+  constant_parameters <- NULL
+
+  if (constant_numerator) {
+    constant_parameters <- c(constant_parameters, numerator_code)
+
+    parameters <- parameters %>%
+      .[!. == numerator_code]
+  }
+
+  if (constant_denominator) {
+    constant_parameters <- c(constant_parameters, denominator_code)
+
+    parameters <- parameters %>%
+      .[!. == denominator_code] %>%
+      ifelse(length(.) == 0, NULL, .)
+  }
```

Comment:
Thanks for the suggested change. Agree to update.

> See my other suggestion about changing how we handle constants. This could perhaps eliminate the need for this code.

This code is needed here because `derive_param_waisthgt` doesn't have `constant_parameters` argument at all where I omitted it on purpose as only HEIGHT can be constant (in line with `admiral::derive_param_bmi`). So we either identify constant parameters here or in every single wrapper separately to build `constant_parameters` argument for `derive_param_computed`. I'd better keep it in `derive_param_ratio`.